python/tests/test_extracted_datatypes.py

```diff
@@ -48,39 +90,89 @@ def sample_input_d(ceiling_input: reader.InputData,
         ],
         'floors': [
             floor_input
-        ]
+        ],
+        'walls': [
+            wall_input
+        ],
+        'codes': codes
+
     }
 
 
 class TestCeiling:
 
     @pytest.fixture
-    def sample(self, sample_input_d: reader.InputData):
+    def sample(self, sample_input_d: reader.InputData) -> typing.Dict[str, str]:
         ceiling = sample_input_d['ceilings'][0]
         ceiling['nominalRsi'] = float(ceiling['nominalRsi'])
         ceiling['effectiveRsi'] = float(ceiling['effectiveRsi'])
         ceiling['area'] = float(ceiling['area'])
         ceiling['length'] = float(ceiling['length'])
         return ceiling
 
-    def test_from_data(self, sample):
+    def test_from_data(self, sample: typing.Dict[str, typing.Any]) -> None:
         output = extracted_datatypes.Ceiling.from_data(sample)
         assert output.label == 'Main attic'
         assert output.area_metres == 46.4515
 
+    def test_to_dict(self, sample: typing.Dict[str, typing.Any]) -> None:
+        output = extracted_datatypes.Ceiling.from_data(sample).to_dict()
+        assert output['areaMetres'] == sample['area']
+
 
 class TestFloor:
 
     @pytest.fixture
-    def sample(self, sample_input_d: reader.InputData):
+    def sample(self, sample_input_d: reader.InputData) -> typing.Dict[str, typing.Any]:
         floor = sample_input_d['floors'][0]
         floor['nominalRsi'] = float(floor['nominalRsi'])
         floor['effectiveRsi'] = float(floor['effectiveRsi'])
         floor['area'] = float(floor['area'])
         floor['length'] = float(floor['length'])
         return floor
 
-    def test_from_data(self, sample):
+    def test_from_data(self, sample: typing.Dict[str, typing.Any]) -> None:
         output = extracted_datatypes.Floor.from_data(sample)
         assert output.label == 'Rm over garage'
         assert output.area_metres == 9.2903
+
+    def test_to_dict(self, sample: typing.Dict[str, typing.Any]) -> None:
+        output = extracted_datatypes.Floor.from_data(sample).to_dict()
+        assert output['areaMetres'] == sample['area']
+
+
+class TestWall:
+
+    @pytest.fixture
+    def sample(self, sample_input_d: reader.InputData) -> typing.Dict[str, typing.Any]:
+        wall = sample_input_d['walls'][0]
+        wall['nominalRsi'] = float(wall['nominalRsi'])
+        wall['effectiveRsi'] = float(wall['effectiveRsi'])
+        wall['perimeter'] = float(wall['perimeter'])
+        wall['height'] = float(wall['height'])
+        return wall
+
+    def test_from_data(self,
+                       sample: typing.Dict[str, typing.Any],
+                       dict_codes: typing.Dict[str, typing.Dict[str, typing.Dict[str, str]]]) -> None:
+        output = extracted_datatypes.Wall.from_data(sample, dict_codes['wall'])
```

Comment:
No, this still doesn't feel right. Interface is too complicated, requiring a special dict construction that only `Wall`  knows about, but doesn't help you to build.

How about making `from_data` an instance method, instead of a class method? This would allow you to pass `codes` (raw from the XML) into the `__init__` method, where it would be transformed into an internal representation that no-one else needs to care about. It could also be re-used across all the calls to `from_data`.

Reply:
That sounds like a reasonable alternative, I'll add that in